pkg/distributor/distributor.go

```diff
@@ -1290,6 +1296,23 @@ func (d *Distributor) cleanupAfterPushFinished(rs *requestState) {
 	}
 }
 
+// checkStartedMiddleware blocks until the underlying distributor service is started.
+func (d *Distributor) checkStartedMiddleware(next PushFunc) PushFunc {
+	return func(ctx context.Context, pushReq *Request) error {
+		next, maybeCleanup := NextOrCleanup(next, pushReq)
+		defer maybeCleanup()
+
+		if d.isServiceStarted.Load() {
+			return next(ctx, pushReq)
+		}
+
+		if err := d.AwaitRunning(ctx); err != nil {
```

Comment:
Should we block push request, or reject it? By blocking here, we can pile up enough requests in the memory that distributor can already crash.

Reply:
Switched to reject in https://github.com/grafana/mimir/pull/9317/commits/010354af2db1e43763209e4c0ce8689a8c129991